File: ReverseLinkedListII.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(head == null){
            return null;
        }
        ListNode currentNode = head;
        ListNode previousNode = null;
        while(left > 1){
            previousNode = currentNode;
            currentNode = currentNode.next;
            left--;
            right--;
        }
        ListNode connection = previousNode;
        ListNode tail = currentNode;
        ListNode prev = null;
        
        while(right-- > 0){
            ListNode temp = currentNode.next;
            currentNode.next = prev;
            prev = currentNode;
            currentNode = temp;
        }
        
        if(connection != null){
            connection.next = prev;
        }else{
            head = prev;
        }
        
        tail.next = currentNode;
        
        return head;
        
    }
}

File: ReverseLinkedListII.js
* Definition for singly-linked list.
 /*
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function(head, left, right) {
    let previousNode = null;
    let traverseNode = head;
    while(left > 1) {
        previousNode = traverseNode;
        traverseNode = traverseNode.next;
        left--;
        right--;
    }
    let connection = previousNode;
    let tail = traverseNode;
    let prev = null;
    while(right > 0) {
        const temp = traverseNode.next;
        traverseNode.next = prev;
        prev = traverseNode;
        traverseNode = temp;
        right--;
    }
    if(connection) connection.next = prev;
    else head = prev;
    tail.next = traverseNode;
    return head;
};
